### ISSUE TITLE
GPTDataset generate sequences longer than expected.

### ISSUE THREAD

**Issue by torshie** [NVIDIA/Megatron-LM#399]
I applied the following patch to print sample lengths, found all samples generated in the `else` branch are of length 2049 when `--seq-length 2048` command line option is specified.

```diff
diff --git a/megatron/data/gpt_dataset.py b/megatron/data/gpt_dataset.py
index 2662b5f..e44ba41 100644
--- a/megatron/data/gpt_dataset.py
+++ b/megatron/data/gpt_dataset.py
@@ -9,6 +9,7 @@ import time
 import numpy as np
 import torch

+import megatron
 from megatron import print_rank_0
 from megatron.core import mpu
 from megatron.data.blendable_dataset import BlendableDataset
@@ -299,6 +300,7 @@ class GPTDataset(torch.utils.data.Dataset):
                 self.doc_idx[doc_index_l],
                 length=offset_l + 1))
             sample = np.concatenate(sample_list)
+            megatron.print_rank_0(f'shape: {sample.shape}')

         if self.return_doc_ids: # for retro preprocessing
             return {'text': np.array(sample, dtype=np.int64),
```

Is this an expected behavior or a bug ?

**Comment by torshie** [comment]
The following patch would fix this, could someone please review it ?

```diff
diff --git a/megatron/data/gpt_dataset.py b/megatron/data/gpt_dataset.py
index 2662b5f..d992dee 100644
--- a/megatron/data/gpt_dataset.py
+++ b/megatron/data/gpt_dataset.py
@@ -262,6 +262,7 @@ class GPTDataset(torch.utils.data.Dataset):
                                   documents, self.indexed_dataset.sizes,
                                   splits_string, num_samples, seq_length, seed,
                                   data_cache_path=data_cache_path)
+        self.seq_length = seq_length


     def __len__(self):
@@ -283,7 +284,7 @@ class GPTDataset(torch.utils.data.Dataset):
             doc_ids.append(self.doc_idx[doc_index_f])
             sample = self.indexed_dataset.get(self.doc_idx[doc_index_f],
                                               offset=offset_f,
-                                              length=offset_l - offset_f + 1)
+                                              length=offset_l - offset_f)
         else:
             # Otherwise, get the rest of the initial document.
             doc_ids.append(self.doc_idx[doc_index_f])
@@ -297,9 +298,10 @@ class GPTDataset(torch.utils.data.Dataset):
             doc_ids.append(self.doc_idx[doc_index_l])
             sample_list.append(self.indexed_dataset.get(
                 self.doc_idx[doc_index_l],
-                length=offset_l + 1))
+                length=offset_l))
             sample = np.concatenate(sample_list)

+        assert sample.shape[0] == self.seq_length
         if self.return_doc_ids: # for retro preprocessing
             return {'text': np.array(sample, dtype=np.int64),
                     'doc_ids': np.array(doc_ids, dtype=np.int64)}
diff --git a/megatron/data/helpers.cpp b/megatron/data/helpers.cpp
index 5c3a054..bd55d88 100644
--- a/megatron/data/helpers.cpp
+++ b/megatron/data/helpers.cpp
@@ -127,24 +127,21 @@ py::array build_sample_idx(const py::array_t<int32_t>& sizes_,

     while (sample_index <= num_samples) {
         // Start with a fresh sequence.
-      int64_t remaining_seq_length = seq_length + 1;
+      int64_t remaining_seq_length = seq_length;
       while (remaining_seq_length != 0) {
             // Get the document length.
 	auto doc_id = static_cast<int64_t>(doc_idx[doc_idx_index]);
 	auto doc_length = static_cast<int64_t>(sizes[doc_id]) - doc_offset;
-	// And add it to the current sequence.
-	remaining_seq_length -= doc_length;
 	// If we have more than a full sequence, adjust offset and set
 	// remaining length to zero so we return from the while loop.
-	// Note that -1 here is for the same reason we have -1 in
-	// `_num_epochs` calculations.
-	if (remaining_seq_length <= 0) {
-	  doc_offset += (remaining_seq_length + doc_length - 1);
+	if (remaining_seq_length <= doc_length) {
+	  doc_offset += remaining_seq_length;
 	  remaining_seq_length = 0;
 	} else {
 	  // Otherwise, start from the begining of the next document.
 	  ++doc_idx_index;
 	  doc_offset = 0;
+	  remaining_seq_length -= doc_length;
 	}
       }
       // Record the sequence.
```

**Comment by lmcafee-nvidia** [comment]
The extra token is used as the label when creating the batch, which you can see here: https://github.com/NVIDIA/Megatron-LM/blob/040eac9414ccbd1301ae16369c3044c5632b7e14/pretrain_gpt.py#L53
